jetty-servlet/src/test/java/org/eclipse/jetty/servlet/ServletUpgradeTest.java

```diff
@@ -0,0 +1,352 @@
+                LOG.info("receivedThirdMessage : " + receivedThirdMessage);
+                if (receivedFirstMessage && receivedSecondMessage && receivedThirdMessage)
+                {
+                    break;
+                }
+            }
+        }
+        finally
+        {
+            try
+            {
+                if (input != null)
+                {
+                    LOG.info("Closing input...");
+                    input.close();
+                    LOG.info("Input closed.");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close input:" + ex.getMessage(), ex);
+            }
+
+            try
+            {
+                if (output != null)
+                {
+                    LOG.info("Closing output...");
+                    output.close();
+                    LOG.info("Output closed .");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close output:" + ex.getMessage(), ex);
+            }
+
+            try
+            {
+                if (s != null)
+                {
+                    LOG.info("Closing socket..." + CRLF);
+                    s.close();
+                    LOG.info("Socked closed.");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close socket:" + ex.getMessage(), ex);
+            }
+        }
+
+        assertTrue(passed1 && passed2 && passed3);
+    }
+
+    private static class TestServlet extends HttpServlet
+    {
+        public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException
+        {
+            if (request.getHeader("Upgrade") != null)
+            {
+                response.setStatus(101);
+                response.setHeader("Upgrade", "YES");
+                response.setHeader("Connection", "Upgrade");
+                request.upgrade(TestHttpUpgradeHandler.class);
+            }
+            else
+            {
+                response.getWriter().println("No upgrade");
+                response.getWriter().println("End of Test");
+            }
+        }
+    }
+
+    public static class TestHttpUpgradeHandler implements HttpUpgradeHandler
+    {
+        public TestHttpUpgradeHandler()
+        {
+        }
+
+        @Override
+        public void destroy()
+        {
+            LOG.debug("===============destroy");
+        }
+
+        @Override
+        public void init(WebConnection wc)
+        {
+            try
+            {
+                ServletInputStream input = wc.getInputStream();
+                ServletOutputStream output = wc.getOutputStream();
+                TestReadListener readListener = new TestReadListener("/", input, output);
+                input.setReadListener(readListener);
+                output.println("===============TCKHttpUpgradeHandler.init");
```

Comment:
You've made the effort to test non-blocking reads, I would go the extra mile and verify that non-blocking writes are possible -- should just be few more lines to add a `WriteListener`.

Reply:
Unfortunately, this isn't as simple as adding a few extra lines to add a `WriteListener`: the `ReadListener` writes to the output so I would need to add some sort of state machine to replicate this behavior.

That would modify this test quite substantially which I would like to avoid as it is almost a 1-1 replica of the TCK test.